test/integration/controller/controller_test.go

```diff
@@ -0,0 +1,721 @@
+	volumeID, err := getVolumeID(volumeName)
+	if len(volumeID) == 0 || err != nil {
+		return false, status.Errorf(codes.Internal, "failed to get volumeID with volumeName (%s)", volumeName)
+	}
+
+	diskName, err := azureutils.GetDiskNameFromAzureManagedDiskURI(volumeID)
+	if len(diskName) == 0 || err != nil {
+		return false, status.Errorf(codes.Internal, "failed to extract diskName from diskURI (%s)", volumeID)
+	}
+
+	labelSelector := fmt.Sprintf("%s=%s", azureutils.VolumeNameLabel, diskName)
+	azVAs, err := azVolumeClient.DiskV1alpha1().AzVolumeAttachments(namespace).List(context.Background(), metav1.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		return false, status.Errorf(codes.Internal, "failed to get AzVolumeAttachment List for volumeName (%s): %v", diskName, err)
+	}
+
+	numReplicas := 0
+	for _, azVA := range azVAs.Items {
+		if azVA.Spec.RequestedRole == v1alpha1.ReplicaRole &&
+			azVA.Status.Detail != nil &&
+			azVA.Status.Detail.PublishContext != nil &&
+			azVA.Status.Detail.Role == v1alpha1.ReplicaRole {
+			numReplicas++
+		}
+	}
+	return numReplicas == desiredNumReplica, nil
+}
+
+func TestAzVolume(t *testing.T) {
+	if !*useDriverV2 {
+		t.Skip("Skipping controller test... supported only for v2 driver")
+	}
+
+	err := testBootstrap()
+	require.NoError(t, err)
+
+	tests := []struct {
+		testName            string
+		volumeName          string
+		capacityRange       *v1alpha1.CapacityRange
+		volumeCapabilities  []v1alpha1.VolumeCapability
+		parameters          map[string]string
+		secrets             map[string]string
+		volumeContentSource *v1alpha1.ContentVolumeSource
+		accessibilityReq    *v1alpha1.TopologyRequirement
+		testFunc            func(*testing.T, string)
+		cleanUpFunc         func(*testing.T, string)
+	}{
+		{
+			testName:   "create volume and check if finalizers have been properly populated.",
+			volumeName: "test-volume",
+			capacityRange: &v1alpha1.CapacityRange{
+				RequiredBytes: 1073741824,
+			},
+			volumeCapabilities: []v1alpha1.VolumeCapability{},
+			parameters: map[string]string{
+				"kind":      "managed",
+				"maxShares": "1",
+				"skuName":   "StandardSSD_LRS",
+			},
+			secrets:             map[string]string{},
+			volumeContentSource: nil,
+			accessibilityReq:    nil,
+			testFunc: func(t *testing.T, volumeName string) {
+				azVolume, err := azVolumeClient.DiskV1alpha1().AzVolumes(namespace).Get(context.Background(), volumeName, metav1.GetOptions{})
+				require.NoError(t, err)
+				require.NotNil(t, azVolume.Finalizers)
+
+				// check if finalizer has been properly populated
+				finalizerFound := false
+				for _, finalizer := range azVolume.Finalizers {
+					finalizerFound = finalizerFound || finalizer == azureutils.AzVolumeFinalizer
+				}
+				require.True(t, finalizerFound)
+			},
+			cleanUpFunc: func(t *testing.T, volumeName string) {
+				// delete volume
+				volumeID, err := getVolumeID(volumeName)
+				require.NotEmpty(t, volumeID)
```

Comment:
Using `require` within the `cleanUpFunc` will prevent the function from completely running if there is a failure and fail the test even if the `testFunc` succeeded. Is this intentional?

Reply:
This was intended as clean up cannot be done without a volumeID.